### DISCUSSION TITLE
Add support for "dynamic value" in permissions/filters/presets

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#2993] [Feature Requests]
We currently support `$CURRENT_USER` / `$CURRENT_ROLE` / `$NOW`, but it would be super useful to be able to use the value of any of the other fields in the check, this would allow you to add validation checks like "Value has to be greater than the value of another field" ("start date has to be before end date")

This could be a `$VALUE()` parameter:

```json
{
    "start_date": {
        "_lt": "$VALUE(end_date)"
    },
    "end_date": {
        "_gt": "$VALUE(start_date)"
    }
}
```

**Comment by Michael-Nussbaumer** [comment]
@rijkvanzanten I was looking for kind of that feature right now. (filter states by selected country and some similar stuff) Is it planed that the `filters/validation/presets` are also available at field settings and not just on permission level.

**Comment by apstanisic** [comment]
Laravel has string based validation that looks simple. With this syntax we can easily have more then one rule per field, without having to use AND and OR

```php
 [
    'title' => 'required|unique:posts|max:255',
    'body' => 'required',
];
```
For example:
```typescript
{
    "start_date": "datetime|required|time_before(end_date)",
    "end_date": "datetime|required|time_after(start_date)",
}

**Comment by WoLfulus** [comment]
String interpolations are kinda fragile IMHO isn't it? Given that the input can come from an end user. We can't just assume we'd never store an actual `$VALUE(end_date)` value in the database itself or anything similar.

This might increase the complexity a little bit when writing them, but what about:

```json
{
    "start_date": {
        "_lt": {
          "value": "some actual value"
        }
    },
    "end_date": {
        "_gt": {
          "field": "start_date"
        }
    }
}
```

Ofc `{ field + value }` should throw in this case if we don't want an additional field just to specify what kind of value we're passing like 
```json
 {
    "_lt": {
      "type": "field",
      "value": "start_date"
    },
    "_gt": {
      "type": "value",
      "value": "some_actual_value"
    }
}
```

I just wonder about JSON columns if we have a "fallback" in place to allow for `"_lt": "value"`

**Reply by rijkvanzanten** [reply]
> String interpolations are kinda fragile IMHO isn't it? Given that the input can come from an end user. We can't just assume we'd never store an actual $VALUE(end_date) value in the database itself or anything similar.

This is theoretically true, but the actual chances of conflict are insignificant compared to the DX / maintenance overhead of splitting it up. We already do it for things like `$CURRENT_USER` / `$CURRENT_ROLE` etc, this discussion is about extending that with a `$VALUE(field_name)`

**Reply by ghost** [reply]
I agree splitting it up would make it too complex. Maybe consider prefixing them with `_directus` or similar, since that is almost never likely to be used by a legitimate user - and throw an error if actually is used for security. Or have a look at my other suggestion below.

**Comment by ghost** [comment]
Another way to do it is to introduce new "column operators" such as "_ceq" (analogous to "_eq", but comparing with a column's value). 

The example would be:
```json
{
    "start_date": {
        "_clt": "end_date"
    },
    "end_date": {
        "_cgt": "start_date"
    }
}

```

This makes sense since they are new operations. This is the approach used by a certain other famous database mirror.

**Comment by devjabeer** [comment]
Any chance of implementing more dynamic values?

**Reply by benhaynes** [reply]
Definitely! List any ideas you have here, and of course, any PRs would be welcomed!

**Comment by Guiqft** [comment]
I'm looking for that too. I wanna use custom permissions like:
On the collection `Companies`:
```json 
{ 
  name:  { 
    "_eq": "$CURRENT_USER.company"
  } 
}

**Reply by Enhed** [reply]
same, how to write filter with access by account field which present in all tables?

**Reply by mlbiche** [reply]
Looking for the same feature 👍

**Comment by Guiqft** [comment]
any chance that this feature will come up? I'm needing more and more and seems that isn't only me

**Reply by rijkvanzanten** [reply]
Sure thing! It's a much requested / wanted feature. That being said, so is everything else under issues currently 🙂 

If you want this feature to land sooner, consider opening a PR or [maybe sponsoring the feature](https://docs.directus.io/getting-started/support/#commissioned-features)

**Reply by Guiqft** [reply]
Oh, of course. I did not want to hurry you, sorry for that.
Just trying to understand the priority of the feature requests, since this discussion has a lot of "upvotes". 
I'm looking to contribute for the app too, since it's being my main work tool at the moment.
Thank you for the answer!

**Comment by Dominic-Marcelino** [comment]
Is this still under consideration? I think it could also solve a lot of SAAS problems like #7054

**Reply by rijkvanzanten** [reply]
Yes certainly! That being said, it's a difficult one to figure out technically, as not every piece of data you would want to add to the permission is available at the time of doing the permission/validation check. For example, currently only user-id and role-id are available, because that information is available globally at the time a request is made. Any other field on user (or any other collection for that matter) would have to dynamically be pulled in right before the permissions are checked. This could cause quite a bit of perf overhead depending on the complexity of the configured permissions, and the size of the payload that's requested 🤔

**Comment by githlc** [comment]
Another idea. Can a handler be opened up, to which some context can also be forwarded? e.g.,

```
{
    "start_date": {
        "_function": "checkStartDate"
    },
    "end_date": {
        "_function": "checkEndDate"
    }
}
```

**Reply by haidv1992** [reply]
@rijkvanzanten 
When does this function avali?

**Comment by NickSettler** [comment]
Can this feature also be added to the Conditions when setting a field of a collection?

**Reply by rijkvanzanten** [reply]
That would be ideal, yeah 👍🏻

**Comment by turbolaserhawk** [comment]
If I could add two cents to this feature, I think it would be ideal if a condition could be dynamically linked to the current value of another sibling input in the same form, with an event handler to change the state of an input if the sibling changes.  

Granted, I think this is not particularly applicable to the API especially if used in an external SPA because multiple calls solve the issue, but I think it's highly desirable in the Directus GUI much in the same way conditional fields were announced as "the most wanted feature" in a recent twitter post. It's a highly functional piece of UI. 

This would allow (for instance), to select a customer, and then a contact belonging to that customer, in a basic CRM or ecommerce admin panel. I am not seeing any other way for an "orders" table to identify which customer placed the order, and them whom in the customer's company.   In a perfect world, I could set the condition on the customer contact input where  $THIS.CustomerID

If not through this feature, is such a thing possible?  Potentially willing to sponsor...

![image](https://user-images.githubusercontent.com/60415816/144725806-8552deb7-0b68-49ef-a72b-3f416da9af7d.png)

**Reply by geekyayush** [reply]
I need this feature please.

**Reply by snimavat** [reply]
I need some thing similar, btw payload cms has support for this

**Comment by claytongulick** [comment]
Can this issue be closed?

At least with filters, it seems to be working with mustache syntax: {{ ... }}

<img width="1440" alt="Screen Shot 2022-06-29 at 1 43 16 PM" src="https://user-images.githubusercontent.com/695284/176512212-f3ab110e-eb23-4035-8117-b047d0715613.png">

**Reply by rijkvanzanten** [reply]
Not _really_. That's a one-off implementation that's only available in the relational interface. I'd like to make sure this functionality is available wherever you're using filter rules 🙂

**Reply by colemahatmccreary** [reply]
Is this relational interface feature not yet available in 12.2.2 (cloud)? Just tried, but can't get it to work with the syntax from the screenshot.

**Reply by rijkvanzanten** [reply]
It is, if you use the many to one interface, the above example will work 👍🏻 That being said, the same syntax isn't consistently available everywhere, so your mileage may vary

**Reply by colemahatmccreary** [reply]
yep, I believe it's the exact same. consider the following model:

model_a
- id
- b_id (fk to model_b)
- c_id (fk to model_c)

model_b
- id (other fields not relevant)

model c
- id
- b_id (fk to model_b)

The c_id field's interface on model_a should only display records where model_c.b_id matches the current record's model_a.b_id field. 

When I save the filter with curly braces syntax it reverts the value as empty/null (see screenshot of the filter editor post-save).

Edit: for the sake of matching the example model, the screenshot's "Crate" field is model_c's b_id field and i'm attempting to save {{b_id}}.

<img width="857" alt="Screen Shot 2022-06-30 at 2 27 38 PM" src="https://user-images.githubusercontent.com/40645344/176781828-e777bf1e-53f3-448a-90df-789ad6a272c5.png">

**Reply by ivanviskovic** [reply]
What i did was, after this step where you selected Crate Equals
Click on the name **Filter** -> Edit Raw Value and then enter "{{b_id}}" after "eq". It should look something like this `"_eq": "{{ b_id }}"`

**Reply by Arecsu** [reply]
Really nice that this exist. Didn't know that it was possible, at least in relational interfaces. Would be cool if it was implemented everywhere!

**Reply by keithotto** [reply]
Does this work for filtering by values in another related field? For eg. could i have `"_eq": "{{ b.related.id }}"`
I tried something similar to that but get an error message when I try to actually add an item to that field.

**Reply by Triggsy22** [reply]
Its quite nice that this is possible, it was exactly what i was searching for.
It should be nice if it would be implemented into the documentation: https://docs.directus.io/reference/filter-rules.html
And also it would be nice if this feature will be implemented in the directus web interface.

**Reply by LeonManolo** [reply]
Does it only work with m2o relations? I can't seem to get it to work with a m2m relation.

**Reply by jclaveau** [reply]
Thanks for the trick! 

For those struggling with `empty` values returned by `micromustache`
![image](https://github.com/directus/directus/assets/1556489/09d82e0d-e770-4da4-aa24-34cd06cf001f)

![image](https://github.com/directus/directus/assets/1556489/e8e36f6d-b426-4dfe-a5e4-b20e77aeaf65)

I found a workaround to provide a default value:
![image](https://github.com/directus/directus/assets/1556489/f28326bf-531e-44ee-8bdf-19ecd0cb3ce6)

Hoping this could help!

**Reply by ritterspach** [reply]
> Does this work for filtering by values in another related field? For eg. could i have "_eq": "{{ b.related.id }}"
I tried something similar to that but get an error message when I try to actually add an item to that field.

@keithotto I have the same challenge, did you found a solution for this?

**Reply by Triggsy22** [reply]
This does not work with all kind of filters.
As an Example: 'Less than' does not work if both values are Decimals.
![image](https://github.com/directus/directus/assets/33725296/c19b2409-da3a-467f-97b5-e069eba5e682)
![image](https://github.com/directus/directus/assets/33725296/1ac70e59-b5de-4822-a937-63fd304b4503)

**Comment by w0kyj** [comment]
Related PR: #14887

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
Should Have:
- top level field support

Could Have:
- support relational fields -- if too complex or exceeds scope then add as separate Feature

**Comment by bernatvadell** [comment]
Hi @rijkvanzanten

Do you know if progress has been made with this feature?

Open this ticket. I think that implementing this improvement could help me.
https://github.com/directus/directus/issues/20459

I have seen some PR that implemented this functionality but for some reason it did not end up being merged into the main branch.
I would be willing to develop this feature if you tell me the minimum requirements.

Thank you

**Reply by infomus** [reply]
Hi, I'm also willing to help, as it would be super helpful for us

**Comment by garytube** [comment]
I need it. Especially for start & end dates

**Reply by ChuckMoe** [reply]
Same here! Simple checks like `start_date + X < end_date` are necessary for us.
Would these conditions also be written into the DB schema? Or would you also inspect the DB schema for conditions like that and apply them?

**Reply by jclaveau** [reply]
It seems that all the Query Api enhancements are stalled until the new data abstraction layer called Durus is shipped. You can follow the stories here: https://github.com/orgs/directus/projects/13/views/12?filterQuery=is%3Aopen+

**Reply by ChuckMoe** [reply]
Thanks! Didn't know about the project.

**Comment by BK-HR** [comment]
Hello,  is there a solution to use the field validation UI message in directus (maybe by flow script or other) to compare the field value to other field ?

**Comment by snimavat** [comment]
How about some thing like Country and States, where I would like to filter/show states, from selected country only.
Bt would be even great, if we can support scenarios like filter cities based on selected country, (Cities are under states and states under country, bt on ui, there's no field for state, just country and city) - This is just an example.  In real we need to do similar thing for product variants,

**Reply by w0kyj** [reply]
There is a Marketplace interface that supports this use case: `Multilevel Autocomplete Interface`

**Reply by snimavat** [reply]
Its not just about autocomplete, when creating/updating records, i need ability to filter data based on some current values in form....

**Comment by turbolaserhawk** [comment]
I agree with this. There are so many obvious applications for passing a dynamic parameter into a filter, and would facilitate many use cases to be able to do so.  

The example given illustrates the need most clearly.  Three collections [ Country, State, City].   Ability when selecting a country to filter the relational list to just the states where State.CountryID  = {$selectedCountry}  , or where City.StateID = {$selectedState}. 

If there is a way to achieve this with current filters, it is not obvious.

**Reply by falmanna** [reply]
Isn't it possible to do it like what's shown in this comment? 
https://github.com/directus/directus/discussions/2993#discussioncomment-3050848

**Comment by FredericLatour** [comment]
I'm a bit at a loss with the various directions of the discussion.
The possibility to filter a field based on another field (i.e.: modified_at > created_at, ...) is important. The necessity to create an endpoint each time such a filter is necessary, is inconvenient.
Is this feature already available, hidden somewhere? Will it be implemented ?

**Reply by rijkvanzanten** [reply]
No, it's not currently available

**Comment by maazamaani** [comment]
I need this too for many projects that lack of deep and dynamic permissions made directus completely useless.. but I want to add more situations that cannot be done using current permissions implemented, but are useful in many technical and business situations.

## Why it is important

As a data platform, its necessary to have full control over data based on various business needs. Today it is a Directus weakness.
many required conditions can be done using api filters, but this results in:
- Data leak in APIs: limited data is not accessible in frontend filtered api call, but is accessible using directus endpoints if user directly call them
- Without a robust permission engine, limited data is accessible through data studio panel. So it is useless for tenants with complex permission needs

## Permission Usecases with Example

template: **`[field] [condition] [value]`**
headlines with ⭐ are very useful in different situations in my experience.

### ⭐  1. Dynamic Value
other fields in this collection as `[value]`
- e.g. **`[date] [_lte] [expire-date]`**

###  ⭐  2. Include Condition
check if a O2M `[field]` includes a `[value]`
- e.g. **`[subscribers] [includes] [$CURRENT_USER]`**

> Update: Thanks to @rijkvanzanten mention, it is possible using `_eq` operation. [More](https://docs.directus.io/reference/filter-rules.html#some-vs-none-in-one-to-many)

###  ⭐  3. Deep Value (is_in)
using deep `[condition]` for o2m and m2m `[field]`s. (set child permissions based on parent value.)
- e.g. [tenance] [is_in] [$CURRENT_USER -> tenances]

### 4. Deep Field (has)
using deep `[condition]` for m2o and m2m `[field]`s. (set parent permissions based on childs value.)
- e.g. **`[blocks -> status] [has] [published]`**

### 5. Nested User Field
using $CURRENT_USER attributes as `[value]`
- e.g. **`[email] [_eq] [$CURRENT_USER -> email]`**

> Update: @rijkvanzanten mentioned that this is possible too with current implementation.

### 6. Aggregated Field
using aggregation and grouping for `[field]`s having relation
- e.g. **`[count(participants)] [_lte] [capacity]`**

### 7. Operations
using mathematical operations in `[value]`.
- e.g. **`[remained] [_gte] [min + 2]`**

**Reply by rijkvanzanten** [reply]
Thanks for the insights 🙂 As far as I understand it, examples 2, 3, and 5 already work currently, as $CURRENT_USER is available there w/ nesting

**Reply by maazamaani** [reply]
You are right.
I double checked and seems example 2 is achievable using `_eq` and example 5 is possible too. but no way for example 3.. this is reproduction steps for example 3:

## No Operation for Example 3

1. I created a collection named `courses` and created a course.

2. Added a m2m field to directus_users, named `courses`

<img width="382" alt="Screenshot 1403-11-11 at 12 37 33" src="https://github.com/user-attachments/assets/3a08acef-bc8b-4f72-8fa7-115348a236d9" />

3. Added a test user and added course 1 to its courses:

<img width="383" alt="Screenshot 1403-11-11 at 12 39 48" src="https://github.com/user-attachments/assets/639912b7-a8f2-42fb-ac4c-2eed58bbc486" />

4. In user policies, added a rule like this:
I know that `_in (is one of)` searchs in an array of object to match entire object and it cannot match inside objects.. but there is no other option for this situation..

```
{
    "_and": [
        {
            "id": {
                "_in": [
                    "$CURRENT_USER.courses"
                ]
            }
        }
    ]
}
```
<img width="489" alt="Screenshot 1403-11-11 at 12 42 19" src="https://github.com/user-attachments/assets/d936a9f0-1bc0-4127-ac38-a908036bf8c9" />

5. In test user's panel, I have no permission to courses. but expected to have course 1 in list.
<img width="388" alt="Screenshot 1403-11-11 at 12 46 45" src="https://github.com/user-attachments/assets/ae9f0b7f-54d9-4e3c-ad59-547ff34a77f6" />

**Comment by snimavat** [comment]
I have a very simple use case.

1. Each category supports certain dynamic fields
2. When a product is being created, and a category is selected, I need to present users with choice of fields only based on the selected product

Currently does not seem possible with directus

**Reply by maazamaani** [reply]
exactly! in following data model, if you want to restrict options in product based on selected category, you should open `product -> options` field setting in data model menu and set filter to [option -> category] [_eq] [product -> category].. but [product -> category] dynamic field is not selectable in filters.

- product:
  - category (m2o)
  - options (m2m)
- options:
  - category (m20)
  - option (string)

**Comment by Vladimir-Va** [comment]
I tried the bypass:
`ProductType=m2o=>Brand=m2o=>Manufacturer`
For see o2m list related ProductType in Manufacturer item.

Calculated Fields Bundle Extension:

- [x] ProductType add Calculated_field and get Manufacturer_ID value
- [x] Create ProductType=m2o=>Manufacturer field and set filter Manufacturer _eq {{ Calculated_field }}
Result: Empty list for selection
**The attempt failed**

https://github.com/judahbsullivan/directus-auto-gen-extension :
When accessing the field m2o every time get `Unexpected Error`
**The attempt failed**
